dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,239 @@
+from __future__ import unicode_literals
+
+import os
+import posixpath
+
+from funcy import cached_property
+from ratelimit import limits, sleep_and_retry
+from backoff import on_exception, expo
+
+from dvc.scheme import Schemes
+from dvc.path_info import CloudURLInfo
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.cached_dirs = {}
+        self.cached_ids = {}
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        if Config.SECTION_REMOTE_KEY_FILE not in self.config:
+            raise DvcException(
+                "Google Drive settings file path is missed from config. "
+                "Learn more at https://dvc.org/doc."
+            )
+        self.gdrive_credentials_path = self.config.get(
+            Config.SECTION_REMOTE_KEY_FILE
+        )
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    @sleep_and_retry
+    @limits(calls=10, period=1)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            if "Rate Limit Exceeded" in str(exception):
+                raise DvcException("API usage rate limit exceeded")
+            raise
+        return result
+
+    def list_drive_item(self, query):
+        list_request = RequestListFilePaginated(self.drive, query)
+        page_list = self.execute_request(list_request)
+        while page_list:
+            for item in page_list:
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            self.cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            self.cached_ids[dir1["id"]] = dir1["title"]
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv("PYDRIVE_USER_CREDENTIALS_DATA"):
+            with open("credentials.json", "w") as credentials_file:
```

Comment:
`credentials.json` is a PYDRIVE-specific name, right? Can we take it as a constant somewhere?

since we are writing the file how can we be sure that we have permissions to write it? 

what happens if file already exists?

do we cleanup it after we done? basically, let's imagine I run it second time and now I don't use `PYDRIVE_USER_CREDENTIALS_DATA`

Reply:
> credentials.json is a PYDRIVE-specific name, right? Can we take it as a constant somewhere?

Sounds good. Yes, it's PYDRIVE-specific, but still configurable by DVC's users.

> since we are writing the file how can we be sure that we have permissions to write it?
what happens if file already exists?
do we cleanup it after we done? basically, let's imagine I run it second time and now I don't use PYDRIVE_USER_CREDENTIALS_DATA

Path to credentials is configurable via `settings.yaml` ( `save_credentials_file: credentials.json` ) and can be fully customized by user. It should be described in the docs, but user is responsible to specify accessible file path. PyDrive loads credentials from specified file path if the file exists. Otherwise PyDrive starts command line sign-in auth process.

Reply:
https://pythonhosted.org/PyDrive/oauth.html#sample-settings-yaml

Reply:
got it. I still think if we provide an interface via env variable we should write credentials into some temp file, clean it up command is done. Basically, this will guarantee that if I run w/o that env variable set it should fail.

Reply:
sounds good, will think about implementation